src/python/pants/backend/python/subsystems/pex_bootstrap/_ptex_launcher.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# This bootstrapping technique for lazily loading requirements comes from the pex project test case:
+# https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+
+import json
+import os
+import sys
+
+from pex import resolver
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+from pex.platforms import Platform
+
+
+self = sys.argv[0]
+filename_base, ext = tuple(os.path.splitext(self))
+hydrated_ipex_file = '{}.ipex'.format(filename_base)
+
+if not os.path.isfile(hydrated_ipex_file):
+  print('Hydrating {} to {}'.format(self, hydrated_ipex_file))
+
+  wrapper_ptex_info = PexInfo.from_pex(self)
+  if wrapper_ptex_info.zip_safe:
+    raise ValueError("Could not hydrate ptex: ipex info must be marked zip_safe=False. This ipex file's info was:\\n{}".format(wrapper_ptex_info.dump(indent=4)))
+  code_root = os.path.join(wrapper_ptex_info.zip_unsafe_cache, wrapper_ptex_info.code_hash)
+  with open_zip(self) as zf:
+    # Populate the pex with the pinned requirements and distribution names & hashes.
+    ipex_info = PexInfo.from_json(zf.read('IPEX-INFO'))
+    ipex_builder = PEXBuilder(pex_info=ipex_info)
+
+    # Populate the pex with the needed code.
+    ptex_info = json.loads(zf.read('PTEX-INFO').decode('utf-8'))
+    for path in ptex_info['code']:
+      ipex_builder.add_source(os.path.join(code_root, path), path)
+
+    # Get the requirements to resolve.
+    plat = Platform.current()
+    transitive_requirements_by_platform = ptex_info['transitive_requirements_by_platform']
+    if plat.platform not in transitive_requirements_by_platform:
+      raise ValueError("Current platform {} was not found in the 'transitive_requirements' field in PTEX-INFO:\\n{}".format(plat, json.dumps(ptex_info, indent=4)))
+
+  # Perform a fully pinned intransitive resolve to hydrate the install cache (not the
+  # pex!).
+  resolver_settings = ptex_info['resolver_settings']
+  # TODO: Convert .indexes and .find_links into the old .fetchers until pants upgrades to pex 2.0!
+  fetchers = [PyPIFetcher(url) for url in resolver_settings['indexes']]
+  del resolver_settings['indexes']
+  fetchers.extend(Fetcher([url]) for url in resolver_settings['find_links'])
+  del resolver_settings['find_links']
+  resolver_settings['fetchers'] = fetchers
+
+  resolved_distributions = resolver.resolve(
+    requirements=transitive_requirements_by_platform[plat.platform],
+    cache=ipex_info.pex_root,
+    transitive=False,
+    platform=plat.platform,
+    interpreter=PythonInterpreter.get(),
+    **resolver_settings
+  )
+
+  for resolved_dist in resolved_distributions:
```

Comment:
Right now, `.add_distribution()` as we do here *is* adding files -- I couldn't figure out another way to get a runnable `.ipex`. See @jsirois's comment: https://github.com/pantsbuild/pants/pull/8793#discussion_r357111750 -- it looks like there's supposed to be a way to avoid adding these distributions, but I haven't figured it out yet.

Reply:
I guess this isn't a blocker for merging (it's just an extraction-time optimisation, rather than a hard requirement), but let's at least add a TODO saying that this _shouldn't_ be necessary. I seem to remember John's prototype adding distributions in some way with a comment, but looking at https://github.com/jsirois/pex/pull/1/files I can't see it, so maybe I hallucinated it...

Reply:
I also thought I recalled that. A TODO for that sounds great, will do!